### ISSUE TITLE
prod JobTracker can't keep up

### ISSUE THREAD

**Issue by bossie** [Open-EO/openeo-geopyspark-driver#373]
Even with an interval of 2 minutes instead of the original 1 minute, messages keep piling up in the prod JobTracker queue; what is slowing it down and how can we address this?

**Comment by soxofaan** [comment]
Maybe a (relatively) quick win:
`ZkJobRegistry.get_running_jobs` roughly does:
```python
user_ids = self._zk.get_children(self._ongoing())
for user_id in user_ids:
    job_ids = self._zk.get_children(self._ongoing(user_id))
```

the `/ongoing` "folder"  lists around 90 users atm, but most of these have no ongoing jobs, so you are doing tens of needless `get_children` calls for these users.
We could clean up empty user folders, or maybe safer: check the "numChildren"  metadata of a user node before descending into it with `get_children`

**Comment by soxofaan** [comment]
Current state of `/openeo/jobs`: 318 users under `ongoing` and 297 users under `done`

I already pruned `/openeo/dev/jobs`, went
- from 91 users under `ongoing` and 82 under `done`
- to 24 users under `ongoing` and 22 under `done`

**Comment by soxofaan** [comment]
now also pruned empty users under `/openeo/jobs` went
- from 318 users under `ongoing` and 297 users under `done`
- to 88 users under `ongoing` and 86 users under `done`

`ZkJobRegistry.get_running_jobs()`  still takes 0:07:35 :confused:

**Comment by soxofaan** [comment]
hmm apparently user `wig` has 14940  jobs listed under `ongoing` which is basically 98% of all ongoing jobs.
That is what makes `ZkJobRegistry.get_running_jobs()` so slow

**Comment by soxofaan** [comment]
I added some options to the cleaner tool/script:
```
  --user USER           Only clean up for specified user id. Can be specified
                        multiple times (default: None)
  --min-age MIN_AGE     Minimum age in days for jobs to clean up (default: 60)
  --dry-run
  --skip-done           Don't clean up `done` jobs (default: True)
  --skip-ongoing        Don't clean up `ongoing` jobs (default: True)
  --jobs-per-user-limit JOBS_PER_USER_LIMIT
                        Maximum number of jobs per user to consider for
                        deletion. Allows keeping the overall execution time
                        reasonable when there are users with excessive number
                        of jobs. (default: 1000)
```

And now I'm using that to more aggressively clean up the `ongoing` folder of user `wig`, which contains a lot of old jobs that where  created but never started. e.g.:
```
python -m openeogeotrellis.cleaner --user jenkins --user wig --min-age 30 --skip-done  --jobs-per-user-limit=1000
```

Note I'm only cleaning up old jobs under  `ongoing`, assuming these are jobs that were never started and consequently don't have an output folder to clean up as well. That way I only need zookeeper deletes, which I can easily run it locally from my laptop,

**Comment by soxofaan** [comment]
Number of `ongoing` jobs for user `wig` is now down to 1314 (from 14940), that should improve the situation considerably I hope.

I also added this more aggressive cleanup for users `jenkins` and `wig`  to scripts/cleaner-entrypoint.sh in 8b2869453446979f313f0580d2cfce528684f919

**Comment by bossie** [comment]
> I think this ticket can be considered resolved for now.
> 
> One topic to follow up on is getting the logs from the cleaner script runs in Kibana for easier monitoring. I don't think that is the case already @bossie ?

True; it should mimic `async_task` and `job_tracker` by writing to a rolling log file that is picked up by filebeat.
